apps/prairielearn/python/question_phases.py

```diff
@@ -0,0 +1,187 @@
+import base64
+import copy
+import io
+import os
+import pathlib
+import sys
+import time
+from typing import Any, Dict, Literal, Optional, Set, Tuple, TypedDict
+
+import lxml.html
+from check_data import Phase, check_data
+from traverse import traverse_and_execute, traverse_and_replace
+from typing_extensions import assert_never
+
+PYTHON_PATH = pathlib.Path(__file__).parent.parent.resolve()
+CORE_ELEMENTS_PATH = (PYTHON_PATH / "elements").resolve()
+SAVED_PATH = copy.copy(sys.path)
+
+
+class ElementInfo(TypedDict):
+    name: str
+    controller: str
+    type: Literal["core", "course"]
+
+
+class RenderContext(TypedDict):
+    html: str
+    """A string consisting of `question.html` with Mustache templating applied."""
+
+    elements: Dict[str, ElementInfo]
+    """A dict mapping an element name to information about them."""
+
+    element_extensions: Dict[str, Dict[str, Dict]]
+    """A dict mapping an element name to a dict of extensions for that element."""
+
+    course_path: str
+    """The path to the course directory."""
+
+
+def filelike_to_string(filelike: Any) -> str:
+    # if val is None, replace it with empty string
+    if filelike is None:
+        filelike = ""
+
+    # if val is a file-like object, read whatever is inside
+    if isinstance(filelike, io.IOBase):
+        filelike.seek(0)
+        filelike = filelike.read()
+
+    # if val is a string, treat it as utf-8
+    if isinstance(filelike, str):
+        filelike = bytes(filelike, "utf-8")
+
+    # if this next call does not work, it will throw an error, because
+    # the thing returned by file() does not have the correct format
+    return base64.b64encode(filelike).decode()
+
+
+def process(
+    phase: Phase, data: dict, context: RenderContext
+) -> Tuple[Optional[str], Set[str]]:
+    html = context["html"]
+    elements = context["elements"]
+    element_extensions = context["element_extensions"]
+    course_path = context["course_path"]
+
+    # This will track which elements have been processed.
+    processed_elements: Set[str] = set()
+
+    # If we're in the `render` phase, we'll eventually capture the HTML here.
+    # If we're in the `file` phase, we'll capture file data here.
+    # Otherwise, this will remain `None`.
+    result = None
+
+    total_time = 0
```

Comment:
`total_time` and other timing variables aren't used anymore. They should be removed along with the `time` import.

Reply:
Done in 8e121e4.